component/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+        let (_, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, _) = oneshotChannel();
+        Supervisor {
+            irrecoverable_signal: tr_irrecoverable,
+            cancellation_signal: Some(tx_cancellation),
+            join_handle: None,
+            manageable: component,
+        }
+    }
+
+    /// Spawn calls the start function of the Manageable component and runs supervision.
+    pub async fn spawn(mut self) -> Result<(), anyhow::Error> {
+        let (tx_irrecoverable, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, tr_cancellation) = oneshotChannel();
+
+        // this assumes the start_fn is populated
+        let wrapped_handle = self
+            .manageable
+            .start(tx_irrecoverable, tr_cancellation)
+            .await;
+
+        self.irrecoverable_signal = tr_irrecoverable;
+        self.cancellation_signal = Some(tx_cancellation);
+        self.join_handle = Some(wrapped_handle);
+
+        self.run().await
+    }
+
+    /// Run watches continuously for irrecoverable errors or JoinHandle completion.
+    pub async fn run(mut self) -> Result<(), anyhow::Error> {
+        // select statement that listens for the following cases:
+        //
+        // Irrecoverable signal incoming => log, terminate and restart
+        // completion of the task => already terminated, log and restart
+        //
+        // The handle_irrecoverable is run before the existing task gets
+        // cancelled by restart in the case that an irrecoverable signal
+        // was sent to us. This makes resource cleanup possible.
+        loop {
+            tokio::select! {
+                Some(message) = self.irrecoverable_signal.recv() => {
+                    self.manageable.handle_irrecoverable(message)?;
+                    self.restart().await?;
+                },
+
+               // Poll the JoinHandle<O>
+               _result =  self.join_handle.as_mut().unwrap(), if self.join_handle.is_some() => {
+                    // this could be due to an un-caught panic
+                    // we don't have a user-supplied message to log, so we create a generic one
+                    let message = anyhow!("An unexpected shutdown was observed in a component.");
+                    self.manageable.handle_irrecoverable(message)?;
+                    self.restart().await?;
+                }
+            }
+        }
+    }
+
+    async fn restart(&mut self) -> Result<(), anyhow::Error> {
+        // restart
+        let (tx_irrecoverable, tr_irrecoverable) = channel(10);
+        let (tx_cancellation, tr_cancellation) = oneshotChannel();
+
+        // call the start method
+        let wrapped_handle: JoinHandle = self
+            .manageable
+            .start(tx_irrecoverable, tr_cancellation)
+            .await;
+
+        // reset the supervision handles & channel end points
+        // dropping the old cancellation_signal implicitly sends cancellation by closing the channel
```

Comment:
this turns out to be ideal because sending a message on a channel without a receiver that is listening causes an error, so this implicit style frees us of reliance on correct user behavior